fix: some combinations of properties for primitives, GLTF, NFT and  OnPointerEvents for ecs7

Suduck opened the pull request:
## Context

While we were developing the new ECS7, we develop the components quickly to be able to iterate quickly and have a first version where we could start coding some ECS7 sample scenes. Those components weren't tested roughly and we created the components of the shape without the OnPointerDown component.

Thes shapes were working fine with the OnPointerDown component but there were certain combinations of values in the properties of the model that weren't working as expected.

Also, the OnPointerDown event wasn't showing the key assigned as the ECS6.

Since we already have the ECS7 in an internal testing phase, all the components needed to be changed at once to maintain compatibility with the current version

In this PR we are tracking the issues with the combination of certain values ( In primitives, GLTF, and NFT, the key of the OnPointer event not showing and improve the performance reducing the number of colliders needed when multiple OnPointer events are in the same entity.



## What does this PR change?

- Fixes all possible combinations of the shape models. Now `withCollisions`, `isPointerBlocker ` and `visible `work independ. This means that you can combine them as you like and they will work always. Prior to this PR not all combinations were working
- Fixes an issue in ECS6 and ECS7 where primitives with the attributes `withCollisions = true` and `isPointerBlocker = false` where not renderer 
- Fixes the key not being shown when you hover on an entity that has an OnPointerEvent 
- Primitives now have their own rendering Game Object instead of the components (mesh renderer, mesh filter) being added to the entity to match parity with GLTF and NFTs
- NFT prefabs have the box collider adjusted to their mesh ( Only the ECS7 prefabs, NFTs have different prefabs for ECS6 and ECS7 since in ECS7 the error part is instantiated only when they fail, ECS6 instantiated it at the beginning and make them invisible. Performance wise is better to instantiate it only when they fail)
- Now all the OnPointerEvents ( like OnPointerDown and OnPointerUp) share the same collider instead of creating different colliders  
- Implement correct disposing of the Cancellation tokens for NFTS



## What does this PR not pretend?

Implement integration test, since that will be done in another PR. Making an integration test would require lots of changes in other systems and will make this PR more unreadable and bigger

## How to test the changes?

Can't be tested easily

1.- Download the scene https://github.com/decentraland/ecs7-template  (recommended branch:  feat/cube-spawner)
2.- Follow the instructions and start the scene with npm start
3.- Use this URL: http://127.0.0.1:8000/?position=0%2C0&ENABLE_ECS7=&renderer-branch=fix%2Fbox-shape-ecs7&kernel-branch=main&SCENE_DEBUG_PANEL=&realm=v1%7E127.0.0.1%3A8000
4.- Change the code to see if everything works as expected

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

Suduck commented on unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/Events/PointerInputRepresentantion.cs:
```diff
@@ -53,21 +53,21 @@ public void SetData(IParcelScene scene, IDCLEntity entity, bool showFeedback, in
         void IPointerEvent.SetHoverState(bool hoverState)
         {
-            pointerEventHandler.SetFeedbackState(showFeedback, hoverState, button.ToString(), hoverText);
+            pointerEventHandler.SetFeedbackState(showFeedback, hoverState, ((WebInterface.ACTION_BUTTON) button).ToString(), hoverText);
```
((WebInterface.ACTION_BUTTON) button).ToString() <-- This will dissapear when we implement the enum for the buttons

Suduck commented on unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/Utils/ECSComponentsUtils.cs:
```diff
@@ -14,29 +13,37 @@ public static void RemoveMaterialTransition(GameObject go)
         {
-            collider.enabled = withCollisions;
-            collider.gameObject.layer = colliderLayer;
+            foreach (Collider collider in meshesInfo.colliders)
+            {
+                collider.enabled = shouldColliderBeEnabled;
+                collider.gameObject.layer = colliderLayer;
+            }
```
done

Suduck commented on unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/Utils/ECSComponentsUtils.cs:
```diff
@@ -140,32 +137,45 @@ public static void DisposeMeshInfo(MeshesInfo meshesInfo)
+
+    public static void DisposeColliders(List<Collider> collidersList)
+    {
+        foreach (Collider collider in collidersList)
```
done

Suduck commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/Camera/Resources/CameraController.prefab:
```diff
@@ -66,7 +66,7 @@ MonoBehaviour:
     LensShift: {x: 0, y: 0}
     GateFit: 2
-    m_SensorSize: {x: 1.6825095, y: 1}
+    m_SensorSize: {x: 2.3333333, y: 1}
```
We added a new layer to the project, and we need to change the camera to render that layer as well.

pbosio replied:
Why do we need the camera to render that layer? Isn't that layer just for collisions?

Suduck replied:
Because the primitives don't have a  game object, they use the same game object. Also, we are working towards have only 1 collider so we can have the collider and the renderer in the same gameobject

pravusjif replied:
So should we change the camera in this PR as well?

pbosio replied:
It should better to have another game object for the collider, I wouldn't mind if unity let us use bitmasks for layer but it is not the case, so for scalability proposes we should put the collider on it own game object

Suduck commented on unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/PointerEventsController/PointerEventsController.cs:
```diff
@@ -431,6 +431,10 @@ public Ray GetRayFromCamera()
             RaycastHitInfo raycastGlobalLayerHitInfo;
             Ray ray = GetRayFromCamera();
 
+            // Note: This is for hot reload, since we can do a hot reload and the scene may be unloaded
+            if (!worldState.loadedScenes.ContainsKey(worldState.currentSceneId))
+                return;
+            
```
done

Suduck commented on unity-renderer/Assets/Scripts/MainScripts/DCL/DataStore/DataStore_ECS7.cs:
```diff
@@ -11,7 +11,11 @@ public class DataStore_ECS7
         public readonly BaseDictionary<long, GameObject> shapesReady = new BaseDictionary<long, GameObject>();
+        
+        // Those are related to the OnPointerEvents
+        public readonly BaseDictionary<long, List<IPointerInputEvent>> entityEvents = new BaseDictionary<long, List<IPointerInputEvent>>();
```
done

Suduck commented on unity-renderer/ProjectSettings/TagManager.asset:
```diff
@@ -34,7 +34,7 @@ TagManager:
   - FX
   - Skybox
-  - 
+  - CollisionsWithoutPointerEvent
```
This layer is needed when we have Collisions and there is no pointer event blocker. This functionality never worked since it was never implemented, we need this layer to take into account this case

pbosio replied:
What do you mean? Having physical collision without pointer block is currently working fine and it is using `CharacterOnly` layer

Suduck replied:
If I remember correctly, this wasn't working, and changing the layer without knowing the exact usage of `CharacterOnly ` layer can leads to bugs, also the characlyOnly layer doesn't seem like a very accurate description

pbosio replied:
There is a test scene here: https://sdk-test-scenes.decentraland.io/?position=0%2C20&realm=v1%7Esdk-test-scenes.decentraland.io
with a chest and a couple of walls.
there is a wall there that has physical collider and do not block the pointer event

Suduck commented on unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/SphereShape/Handler/ECSSphereShapeHandler.cs:
```diff
@@ -30,9 +30,12 @@ public void OnComponentRemoved(IParcelScene scene, IDCLEntity entity)
 
         public void OnComponentModelUpdated(IParcelScene scene, IDCLEntity entity, PBSphereShape model)
         {
+            if (lastModel != null && lastModel.Equals(model))
```
I don't think so. If the model is exactly the same the component will be already behaving the way you want. This is done because if from the code you use `getMutableGroupOf ` all the components will be updated even if you don't change them

Suduck commented on unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/NFTShape/NFTShapeFrame.cs:
```diff
@@ -100,9 +106,9 @@ private void Awake()
+        public bool HasCollisions() { return meshCollider.enabled; }
         
-        public BoxCollider nftCollider  => boxCollider;
+        public Collider nftCollider  => meshCollider;
```
Because since we are using a mesh collider for the PointerEvent, we need to have the same collider to it doesn't block his self when we are activating IsPointerBlocker

pbosio replied:
not sure if I understand. why would it block itself if there should be colliders using different layers?

Suduck replied:
I will try to explain it better.
Imagine the following situation:

1. We set the model of the NFT to `IsPointerBlocker = true`
2. The `NFTComponentHandler` will change the BoxCollider of the NFT so it blocks the pointer event
3. We add an `OnPointerDown` component to the NFT and it will create a mesh collider from the mesh to handle the event
4. Since the `BoxCollider `is bigger (because it is less accurate) than the mesh collider, it will block the pointer event raycast, so the `MeshCollider ` of the `OnPointerDown ` will never receive the `OnHover ` event since the box collider will block always the raycast

pravusjif replied:
Uhmmm, if that was a problem with the BoxCollider shouldn't we be having this same problem with ECS6 in production today? how are we dealing with that with the current ECS? I'm not sure if we are having that problem with the current ECS6 using a BoxShape, maybe we can test that with a simple test scene and see how it's currently behaving

pbosio replied:
2- We should probably use BoxCollider for physical collision and create a MeshCollider for pointer events (including pointer blocker)
But I didn't check how does that situation is heaving currently on ECS6, cause the answer for this situation is that they should behave the same way, otherwise we are breaking compatibilities.

Suduck replied:
We are having this problem in production today: 
![image](https://user-images.githubusercontent.com/54064092/180981053-bac6d430-09ab-45b6-9fe2-faa87703c3b2.png)

Suduck replied:
> 2- We should probably use BoxCollider for physical collision and create a MeshCollider for pointer events (including pointer blocker) But I didn't check how does that situation is behaving currently on ECS6, cause the answer for this situation is that they should behave the same way, otherwise we are breaking compatibilities.

I don't think that this approach is a good one.  If a NFTShape comes with isPointerBlocker we will have to create 2 colliders even if it doesn't have an OnPointerDown.

We can't make ECS7 based on the compatibility with ECS6 or we will drag the problems of ECS6 into ECS7


The correct approach in my opinion is to have only 1 collider and change the layer taking into account the current status. This is no

pbosio replied:
But NFTShape does block pointer, the reason it's not working on your example is `withCollisions = false` which looks like it's intentionally https://github.com/decentraland/unity-renderer/blob/dev/unity-renderer/Assets/Scripts/MainScripts/DCL/Components/BaseShape/BaseShape.cs#L90-L93
so it doesn't seem to have anything to do with the collider itself.
What I meant with the 2nd point it's that changing behavior it's OK only if it's extensible enough to recreate the ECS6 behavior because we can't break already deployed scenes

Suduck replied:
Yes, but that case can always be done in the ECS6- ECS7 conversor. 

It doesn't have sense that the isPointerBlocker attribute works only if another property is true.  Nico tolds us that one time that he wanted to act just as expected, he needed to add an invisible shape to block the pointer.

I think that we should talk about this with mendez to see what we should do

pravusjif commented on unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/OnPointerEvent.cs:
```diff
@@ -21,11 +21,17 @@ public OnPointerEventHandler()
         {
             this.entity = entity;
-            eventColliders.Initialize(entity);
+            eventColliders.DeprecatedInitialize(entity);
```
Maybe we should leave a `//TODO` here saying that?

pravusjif commented on unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/OnPointerEventColliders.cs:
```diff
@@ -23,8 +23,15 @@ public string GetMeshName(Collider collider)
+            GenerateColliders(entity);
+        }
+
+        // This function should be deleted when delete de old ECS
```
```suggestion
        // This function should be deleted when we delete the old ECS6
```

pravusjif commented on unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/OnPointerEventColliders.cs:
```diff
@@ -137,10 +157,48 @@ void DestroyColliders()
+                onPointerEventGameObjectCollider.layer = PhysicsLayers.onPointerEventLayer; 
+                onPointerEventGameObjectCollider.transform.SetParent(renderers[i].transform, false);
+                
+                // We need to increase it a little bit since if the shape is pointer blocker, it shouldn't block its own pointer event
+                // This way, there is no collision fighting 
+                onPointerEventGameObjectCollider.transform.localScale *= 1.0001f;
```
This looks hacky =S, I think we should check how we are dealing with this situation (having the BoxShape in NFTshapes for example) in the current ECS6 implementation and do the same

Suduck replied:
The current ECS6 implementation is bugged with this. Since the BoxShape in the NFTShapes never blocks the pointer event.  With the current implementation of ECS6, this is the only way to fix it ( in both ECS). 

In my opinion, we should use only 1 collider that acts as everything

Suduck replied:
This has been removed with the current approach

pravusjif commented on unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/CollidersManager/CollidersManager.cs:
```diff
@@ -168,5 +168,35 @@ public void ConfigureColliders(GameObject meshGameObject, bool hasCollision, boo
+            entity.meshesInfo.colliders.Clear(); 
+            
+            // This is de default layer, however, take into account that OnPointerEvents can change this layer
+            int colliderLayer = layerMask;//LayerUtils.CalculateLayerMask(entity.entityId, hasCollision, isPointerBlocker);
```
We should remove the `//LayerUtils.CalculateLayerMask(entity.entityId, hasCollision, isPointerBlocker);` part right?

pravusjif commented:
in ropsten 54,-71 we have a pointer events test scene, if we could re-implement that scene with ECS 7 we could test everything works as expected, and also add the case of an NFTShape with a pointerevents component

pravusjif reviewed (approved):
I tested the build and checked the code.

The main concern would be the case of changing NFTShapes BoxCollider with a MeshCollider and then re-scaling the pointerevents collider, which looks pretty hacky. I think we should check how we are dealing with that situation for the current ECS6 and do the same, but since this doesn't affect production right now I'm approving the PR.

Suduck commented:
> 

It is currently a bug in ECS6.

pbosio commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Components/LoadableShapes/NFTShape/NFTShapeLoaderController.cs:
```diff
@@ -8,7 +8,7 @@
 public interface INFTShapeLoaderController
 {
-    BoxCollider nftCollider { get; }
+    Collider nftCollider { get; }
```
why are we changing this?

Suduck replied:
Since we don't need a specific thing of a BoxCollider, this was done to be more flexible in the future if we want to implement NFTShapes with another collider. Anyway, I have undone it

pbosio commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Configuration/Configuration.cs:
```diff
@@ -216,18 +216,21 @@ public static UnityEngine.Vector2
+
+        // Layers related to the Avatar
+        public static readonly int characterLayer = LayerMask.NameToLayer("CharacterController");
+        public static readonly int collisionsWithoutPointerEvent = LayerMask.NameToLayer("CollisionsWithoutPointerEvent");
```
name shouldn't be `CollisionsWithoutPointerEvent` since a layer represent a bit that will be on or off.
so name should indicate what does that bit is used for

pbosio reviewed (commented):
why are we changing so many prefabs, camera and `DynamicsManager`?
shouldn't we add tests to check if changes are working as intended?
could you also add a little more detailed explanation on the PR description? since is a big PR and by just reading the code it's not clear what was broken, what are we trying to fix and how

Suduck commented on unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/NFTShape/NFTShapeFrame.cs:
```diff
@@ -111,21 +122,22 @@ private void Start()
+        public void SetLayerMask(int layerMask)
         {
-            int colliderLayer = isPointerBlocker ? PhysicsLayers.onPointerEventLayer : DCL.Configuration.PhysicsLayers.defaultLayer;
-
-            boxCollider.gameObject.layer = colliderLayer;
+            boxCollider.gameObject.layer = layerMask;
```
Good catch!

Suduck commented on unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/OnPointerEventColliders.cs:
```diff
@@ -137,10 +157,45 @@ void DestroyColliders()
 
                 if (collider != null)
-                    UnityEngine.Object.Destroy(collider.gameObject);
+                    UnityEngine.Object.Destroy(collider);
```
Good catch! Undone

Suduck commented on unity-renderer/Assets/DCLPlugins/UUIDEventComponentsPlugin/UUIDComponent/OnPointerEventColliders.cs:
```diff
@@ -137,10 +157,45 @@ void DestroyColliders()
+                colliderNames[collider] = meshCollider.sharedMesh.name;
+            else
+                colliderNames[collider] = collider.gameObject.name;
+        }
```
Removed

Suduck commented:
> why are we changing so many prefabs, camera and `DynamicsManager`? shouldn't we add tests to check if changes are working as intended? could you also add a little more detailed explanation on the PR description? since is a big PR and by just reading the code it's not clear what was broken, what are we trying to fix and how

I have added a more detailed description to this PR.

For the DynamicsManager,  since we have renamed the layer `characterOnly`, it has been updated, however, I have reverted the changes and saw no issues so I just revert it

Regarding the camera prefab I have detected a bug in the ECS 6 and also in ECS 7 where primitives were not rendered if they have  `withCollisions = true` and `isPointerBlocker = false`. This was because the layer  `characterOnly ` ( now called `CollisionsWithCharacter` was not rendered. I have added this layer to the camera to solve this problem. Also, I have checked in gensis plaza, casinos and wondermine with a custom script that I made to search for gameobjects with that layer and seems that nothing else is using that layer so it is safe to add it to the rendering of the camera

All the prefabs that are changed are the NFTs for ECS7. Since first I changed them to mesh collider and then I needed to add it back, the prefab changed, when I took a look at the changes, the only thing that changed was the size, now the size is adjusted to the mesh ( unity adjust them automatically if you remove the component and add it again). I thought it was a nice addition and since nobody is working on the ECS7 prefabs its safe to change it

pbosio commented:
> Regarding the camera prefab I have detected a bug in the ECS 6 and also in ECS 7 where primitives were not rendered if they have `withCollisions = true` and `isPointerBlocker = false`. This was because the layer `characterOnly ` ( now called `CollisionsWithCharacter` was not rendered. I have added this layer to the camera to solve this problem. Also, I have checked in gensis plaza, casinos and wondermine with a custom script that I made to search for gameobjects with that layer and seems that nothing else is using that layer so it is safe to add it to the rendering of the camera

Adding a collision layer to the camera rendering is not a proper fix. We shouldn't do it.
If object are not being renderer in ECS6 we can add a ticket and properly tackle it on another PR

> All the prefabs that are changed are the NFTs for ECS7. Since first I changed them to mesh collider and then I needed to add it back, the prefab changed, when I took a look at the changes, the only thing that changed was the size, now the size is adjusted to the mesh ( unity adjust them automatically if you remove the component and add it again). I thought it was a nice addition and since nobody is working on the ECS7 prefabs its safe to change it

I'm not sure if there was a reason for the colliders to be the size they were and we also need to take into account the compatibility layer from ECS6 to ECS7, this change could break some deployed scene's experiences, so we shouldn't be rushing it, we need to do a better assessment before deciding if we should change those colliders or not

Suduck commented:
> > Regarding the camera prefab I have detected a bug in the ECS 6 and also in ECS 7 where primitives were not rendered if they have `withCollisions = true` and `isPointerBlocker = false`. This was because the layer `characterOnly ` ( now called `CollisionsWithCharacter` was not rendered. I have added this layer to the camera to solve this problem. Also, I have checked in gensis plaza, casinos and wondermine with a custom script that I made to search for gameobjects with that layer and seems that nothing else is using that layer so it is safe to add it to the rendering of the camera
> 
> Adding a collision layer to the camera rendering is not a proper fix. We shouldn't do it. If object are not being renderer in ECS6 we can add a ticket and properly tackle it on another PR
> 
> > All the prefabs that are changed are the NFTs for ECS7. Since first I changed them to mesh collider and then I needed to add it back, the prefab changed, when I took a look at the changes, the only thing that changed was the size, now the size is adjusted to the mesh ( unity adjust them automatically if you remove the component and add it again). I thought it was a nice addition and since nobody is working on the ECS7 prefabs its safe to change it
> 
> I'm not sure if there was a reason for the colliders to be the size they were and we also need to take into account the compatibility layer from ECS6 to ECS7, this change could break some deployed scene's experiences, so we shouldn't be rushing it, we need to do a better assessment before deciding if we should change those colliders or not

Ok reverted all the prefabs changes and create a new ticket for the primitives

Suduck commented:
After talking with Pato, he has suggested closing this PR since some changes are going to be done in the GLTF and primitives
